File: MlpNetwork/networkdataset.cpp
#include "stdafx.h"
#include "networkdataset.h"
#include "matrixhelper.h"

namespace mlp_network
{
	// Создаёт пустой новый набор данных для сети.
	NetworkDataset::NetworkDataset() : size_(0), inputSize_(0), outputSize_(0)
	{
	}

	// Создаёт новый набор данных для сети, используя матрицы входных и выходных данных.
	NetworkDataset::NetworkDataset(const matrix<double> &inputData, const matrix<double> &outputData)
		: size_(inputData.size()), inputSize_(inputData[0].size()), outputSize_(outputData[0].size()),
		inputData_(inputData), outputData_(outputData)
	{
	}

	// Создаёт новый набор данных для сети, используя размер набора данных, а также размеры входной и выходной выборки данных.
	NetworkDataset::NetworkDataset(size_t size, size_t inputSize, size_t outputSize)
		: size_(size), inputSize_(inputSize), outputSize_(outputSize),
		inputData_(MatrixHelper::createMatrix<double>(size, inputSize)),
		outputData_(MatrixHelper::createMatrix<double>(size, outputSize))
	{	
	}

	// Возвращает размер набора данных для сети.
	size_t NetworkDataset::size() const
	{
		return size_;
	}

	// Возвращает матрицу входных данных сети.
	const matrix<double>& NetworkDataset::inputData() const
	{
		return inputData_;
	}

	// Задаёт матрицу входных данных сети.
	void NetworkDataset::setInputData(const matrix<double> &inputData)
	{
		inputData_ = inputData;
	}

	// Возвращает матрицу выходных данных сети.
	const matrix<double>& NetworkDataset::outputData() const
	{
		return outputData_;
	}

	// Задаёт матрицу выходных данных сети.
	void NetworkDataset::setOutputData(const matrix<double> &outputData)
	{
		outputData_ = outputData;
	}
}

File: MlpNetworkCudaCSharp/MlpNetwork/NetworkDataSet.cs
using System;

namespace MlpNetwork
{
    public class NetworkDataSet
    {
        private float[][] inputData;
        private float[][] outputData;

        public int NumInput { get; private set; }
        public int NumOutput { get; private set; }
        public int NumSamples { get; private set; }

        public NetworkDataSet(int numInput, int numOutput, int numSamples)
        {
            NumInput = numInput;
            NumOutput = numOutput;
            NumSamples = numSamples;

            inputData = new float[numSamples][];
            outputData = new float[numSamples][];
            for (int i = 0; i < numSamples; i++)
			{
                inputData[i] = new float[numInput];
                outputData[i] = new float[numOutput];
			}
        }

        public NetworkDataSet(float[][] inputData, float[][] outputData)
            : this(inputData[0].Length, outputData[0].Length, inputData.Length)
        {
            for (int i = 0; i < NumSamples; i++)
            {
                Array.Copy(inputData[i], this.inputData[i], NumInput);
                Array.Copy(outputData[i], this.outputData[i], NumOutput);
            }
        }

        public float[][] GetInputData()
        {
            float[][] inputDataCopy = new float[NumSamples][];
            for (int i = 0; i < inputDataCopy.Length; i++)
            {
                inputDataCopy[i] = (float[])this.inputData[i].Clone();
            }

            return inputDataCopy;
        }

        public float[][] GetOutputData()
        {
            float[][] outputDataCopy = new float[NumSamples][];
            for (int i = 0; i < outputDataCopy.Length; i++)
            {
                outputDataCopy[i] = (float[])this.outputData[i].Clone();
            }

            return outputDataCopy;
        }

        public void SetInputData(float[][] inputData)
		{
			if (inputData.Length != NumSamples || inputData[0].Length!= NumInput)
				throw new ArgumentException("Bad number of inputData");

            for (int i = 0; i < inputData.Length; i++)
            {
                Array.Copy(inputData[i], this.inputData[i], NumInput);
            }
		}

		public void SetOutputData(float[][] outputData)
		{
			if (outputData.Length != NumSamples || outputData[0].Length != NumOutput)
                throw new ArgumentException("Bad number of outputData");

            for (int i = 0; i < outputData.Length; i++)
            {
                Array.Copy(outputData[i], this.outputData[i], NumOutput);
            }
		}
    }
}
